### ISSUE TITLE
Rollup v59 breaks application logic

### ISSUE THREAD

**Issue by Havunen** [rollup/rollup#2237]
Hi

Rollup v59+ removes function calls from object literals that are used by external code.

I have created small piece of code to reproduce the problem.
[Rollup v59](https://rollupjs.org/repl?version=0.59.4&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QmNyZWF0ZURldlRvb2xzQnJpZGdlJTdEJTIwZnJvbSUyMCcuJTJGbWF0aHMnJTNCJTVDbiU1Q253aW5kb3cuZGV2dG9vbHMlMjAlM0QlMjBjcmVhdGVEZXZUb29sc0JyaWRnZSgpJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1hdGhzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwY3JlYXRlRGV2VG9vbHNCcmlkZ2UoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjB2YXIlMjBNb3VudCUyMCUzRCUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBfaW5zdGFuY2VzQnlSZWFjdFJvb3RJRCUzQSUyMCU1QiU1RCUyQyU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMkYlMkYlMjB0c2xpbnQlM0FkaXNhYmxlLW5leHQtbGluZSUzQW5vLWVtcHR5JTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMF9yZW5kZXJOZXdSb290Q29tcG9uZW50JTNBJTIwZnVuY3Rpb24lMjBfcmVuZGVyTmV3Um9vdENvbXBvbmVudChpbnN0YW5jZSklMjAlN0IlMjAlN0QlNUNuJTIwJTIwJTIwJTIwJTdEJTNCJTVDbiUyMCUyMCUyMCUyMHZhciUyMFJlY29uY2lsZXIlMjAlM0QlMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTJGJTJGJTIwdHNsaW50JTNBZGlzYWJsZS1uZXh0LWxpbmUlM0Fuby1lbXB0eSU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBtb3VudENvbXBvbmVudCUzQSUyMGZ1bmN0aW9uJTIwbW91bnRDb21wb25lbnQoaW5zdGFuY2UpJTIwJTdCJTIwJTdEJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMHRzbGludCUzQWRpc2FibGUtbmV4dC1saW5lJTNBbm8tZW1wdHklNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcGVyZm9ybVVwZGF0ZUlmTmVjZXNzYXJ5JTNBJTIwZnVuY3Rpb24lMjBwZXJmb3JtVXBkYXRlSWZOZWNlc3NhcnkoaW5zdGFuY2UpJTIwJTdCJTIwJTdEJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMHRzbGludCUzQWRpc2FibGUtbmV4dC1saW5lJTNBbm8tZW1wdHklNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcmVjZWl2ZUNvbXBvbmVudCUzQSUyMGZ1bmN0aW9uJTIwcmVjZWl2ZUNvbXBvbmVudChpbnN0YW5jZSklMjAlN0IlMjAlN0QlMkMlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTJGJTJGJTIwdHNsaW50JTNBZGlzYWJsZS1uZXh0LWxpbmUlM0Fuby1lbXB0eSU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB1bm1vdW50Q29tcG9uZW50JTNBJTIwZnVuY3Rpb24lMjB1bm1vdW50Q29tcG9uZW50KGluc3RhbmNlKSUyMCU3QiUyMCU3RCU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwJTJGKiolMjBOb3RpZnklMjBkZXZ0b29scyUyMHRoYXQlMjBhJTIwbmV3JTIwY29tcG9uZW50JTIwaW5zdGFuY2UlMjBoYXMlMjBiZWVuJTIwbW91bnRlZCUyMGludG8lMjB0aGUlMjBET00uJTIwKiUyRiU1Q24lMjAlMjAlMjAlMjB2YXIlMjBjb21wb25lbnRBZGRlZCUyMCUzRCUyMGZ1bmN0aW9uJTIwKHZOb2RlKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB2YXIlMjBpbnN0YW5jZSUyMCUzRCUyMHVwZGF0ZVJlYWN0Q29tcG9uZW50KHZOb2RlJTJDJTIwbnVsbCklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwaWYlMjAoaXNSb290Vk5vZGUodk5vZGUpKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBpbnN0YW5jZS5fcm9vdElEJTIwJTNEJTIwbmV4dFJvb3RLZXkocm9vdHMpJTNCJTVDbiU1Q3QlNUN0JTVDdCU1Q3QlNUN0JTJGJTJGJTIwQ29tbWVudDElNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcm9vdHMlNUJpbnN0YW5jZS5fcm9vdElEJTVEJTIwJTNEJTIwaW5zdGFuY2UlM0IlNUNuJTVDdCU1Q3QlNUN0JTVDdCU1Q3QlMkYlMkYlMjBUaGlzJTIwY29tbWVudCUyMGlzJTIwYWxzbyUyMHJlbW92ZWQlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwTW91bnQuX3JlbmRlck5ld1Jvb3RDb21wb25lbnQoaW5zdGFuY2UpJTNCJTVDbiU1Q3QlNUN0JTVDdCU1Q3QlNUN0JTJGJTJGJTIwQ29tbWVudDIlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHZpc2l0Tm9uQ29tcG9zaXRlQ2hpbGRyZW4oaW5zdGFuY2UlMkMlMjBmdW5jdGlvbiUyMChjaGlsZEluc3QpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNoaWxkSW5zdC5faW5EZXZUb29scyUyMCUzRCUyMHRydWUlM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwUmVjb25jaWxlci5tb3VudENvbXBvbmVudChjaGlsZEluc3QpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCU3RCklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwUmVjb25jaWxlci5tb3VudENvbXBvbmVudChpbnN0YW5jZSklM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTNCJTVDbiUyMCUyMCUyMCU1Q24lMjAlMjAlMjAlMjByZXR1cm4lMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwQ29tcG9uZW50VHJlZSUzQSUyMENvbXBvbmVudFRyZWUlMkMlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwTW91bnQlM0ElMjBNb3VudCUyQyU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBSZWNvbmNpbGVyJTNBJTIwUmVjb25jaWxlciUyQyU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBjb21wb25lbnRBZGRlZCUzQSUyMGNvbXBvbmVudEFkZGVkJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbXBvbmVudFJlbW92ZWQlM0ElMjBjb21wb25lbnRSZW1vdmVkJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbXBvbmVudFVwZGF0ZWQlM0ElMjBjb21wb25lbnRVcGRhdGVkJTVDbiUyMCUyMCUyMCUyMCU3RCUzQiU1Q24lN0QlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=)

And in v58 this is working correctly [LINK](https://rollupjs.org/repl?version=0.58.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QmNyZWF0ZURldlRvb2xzQnJpZGdlJTdEJTIwZnJvbSUyMCcuJTJGbWF0aHMnJTNCJTVDbiU1Q253aW5kb3cuZGV2dG9vbHMlMjAlM0QlMjBjcmVhdGVEZXZUb29sc0JyaWRnZSgpJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1hdGhzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwY3JlYXRlRGV2VG9vbHNCcmlkZ2UoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjB2YXIlMjBNb3VudCUyMCUzRCUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBfaW5zdGFuY2VzQnlSZWFjdFJvb3RJRCUzQSUyMCU1QiU1RCUyQyU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMkYlMkYlMjB0c2xpbnQlM0FkaXNhYmxlLW5leHQtbGluZSUzQW5vLWVtcHR5JTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMF9yZW5kZXJOZXdSb290Q29tcG9uZW50JTNBJTIwZnVuY3Rpb24lMjBfcmVuZGVyTmV3Um9vdENvbXBvbmVudChpbnN0YW5jZSklMjAlN0IlMjAlN0QlNUNuJTIwJTIwJTIwJTIwJTdEJTNCJTVDbiUyMCUyMCUyMCUyMHZhciUyMFJlY29uY2lsZXIlMjAlM0QlMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTJGJTJGJTIwdHNsaW50JTNBZGlzYWJsZS1uZXh0LWxpbmUlM0Fuby1lbXB0eSU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBtb3VudENvbXBvbmVudCUzQSUyMGZ1bmN0aW9uJTIwbW91bnRDb21wb25lbnQoaW5zdGFuY2UpJTIwJTdCJTIwJTdEJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMHRzbGludCUzQWRpc2FibGUtbmV4dC1saW5lJTNBbm8tZW1wdHklNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcGVyZm9ybVVwZGF0ZUlmTmVjZXNzYXJ5JTNBJTIwZnVuY3Rpb24lMjBwZXJmb3JtVXBkYXRlSWZOZWNlc3NhcnkoaW5zdGFuY2UpJTIwJTdCJTIwJTdEJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMHRzbGludCUzQWRpc2FibGUtbmV4dC1saW5lJTNBbm8tZW1wdHklNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcmVjZWl2ZUNvbXBvbmVudCUzQSUyMGZ1bmN0aW9uJTIwcmVjZWl2ZUNvbXBvbmVudChpbnN0YW5jZSklMjAlN0IlMjAlN0QlMkMlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTJGJTJGJTIwdHNsaW50JTNBZGlzYWJsZS1uZXh0LWxpbmUlM0Fuby1lbXB0eSU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB1bm1vdW50Q29tcG9uZW50JTNBJTIwZnVuY3Rpb24lMjB1bm1vdW50Q29tcG9uZW50KGluc3RhbmNlKSUyMCU3QiUyMCU3RCU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwJTJGKiolMjBOb3RpZnklMjBkZXZ0b29scyUyMHRoYXQlMjBhJTIwbmV3JTIwY29tcG9uZW50JTIwaW5zdGFuY2UlMjBoYXMlMjBiZWVuJTIwbW91bnRlZCUyMGludG8lMjB0aGUlMjBET00uJTIwKiUyRiU1Q24lMjAlMjAlMjAlMjB2YXIlMjBjb21wb25lbnRBZGRlZCUyMCUzRCUyMGZ1bmN0aW9uJTIwKHZOb2RlKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB2YXIlMjBpbnN0YW5jZSUyMCUzRCUyMHVwZGF0ZVJlYWN0Q29tcG9uZW50KHZOb2RlJTJDJTIwbnVsbCklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwaWYlMjAoaXNSb290Vk5vZGUodk5vZGUpKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBpbnN0YW5jZS5fcm9vdElEJTIwJTNEJTIwbmV4dFJvb3RLZXkocm9vdHMpJTNCJTVDbiU1Q3QlNUN0JTVDdCU1Q3QlNUN0JTJGJTJGJTIwQ29tbWVudDElNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcm9vdHMlNUJpbnN0YW5jZS5fcm9vdElEJTVEJTIwJTNEJTIwaW5zdGFuY2UlM0IlNUNuJTVDdCU1Q3QlNUN0JTVDdCU1Q3QlMkYlMkYlMjBUaGlzJTIwY29tbWVudCUyMGlzJTIwYWxzbyUyMHJlbW92ZWQlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwTW91bnQuX3JlbmRlck5ld1Jvb3RDb21wb25lbnQoaW5zdGFuY2UpJTNCJTVDbiU1Q3QlNUN0JTVDdCU1Q3QlNUN0JTJGJTJGJTIwQ29tbWVudDIlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHZpc2l0Tm9uQ29tcG9zaXRlQ2hpbGRyZW4oaW5zdGFuY2UlMkMlMjBmdW5jdGlvbiUyMChjaGlsZEluc3QpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNoaWxkSW5zdC5faW5EZXZUb29scyUyMCUzRCUyMHRydWUlM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwUmVjb25jaWxlci5tb3VudENvbXBvbmVudChjaGlsZEluc3QpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCU3RCklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwUmVjb25jaWxlci5tb3VudENvbXBvbmVudChpbnN0YW5jZSklM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTNCJTVDbiUyMCUyMCUyMCU1Q24lMjAlMjAlMjAlMjByZXR1cm4lMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwQ29tcG9uZW50VHJlZSUzQSUyMENvbXBvbmVudFRyZWUlMkMlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwTW91bnQlM0ElMjBNb3VudCUyQyU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBSZWNvbmNpbGVyJTNBJTIwUmVjb25jaWxlciUyQyU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBjb21wb25lbnRBZGRlZCUzQSUyMGNvbXBvbmVudEFkZGVkJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbXBvbmVudFJlbW92ZWQlM0ElMjBjb21wb25lbnRSZW1vdmVkJTJDJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbXBvbmVudFVwZGF0ZWQlM0ElMjBjb21wb25lbnRVcGRhdGVkJTVDbiUyMCUyMCUyMCUyMCU3RCUzQiU1Q24lN0QlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=)

See how valid code gets removed inside `componentAdded` method

**Comment by guybedford** [comment]
Thanks for posting.

A simpler replication seems to be something like - https://rollupjs.org/repl?version=0.59.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMHglMjAlM0QlMjAlN0IlNUNuJTVDdGZuJTNBJTIwZnVuY3Rpb24lMjBwdXJlJTIwKCklMjAlN0IlN0QlNUNuJTdEJTNCJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHJ1blJuJTIwKCklMjAlN0IlNUNuJTVDdHguZm4oKSUzQiU1Q24lN0QlNUNuJTVDbmV4cG9ydCUyMCU3QiUyMHglMjAlN0QlM0IlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

Basically, we cannot assume we know the definition of an object path if that object is exposed at all either as an entry point export, global, or through an unknown function call parameter.

**Comment by lukastaegert** [comment]
Hi, while I was hiking the last days, this was actually one puzzle piece I was thinking about that we definitely need to fix. What a coincidence. A fix should not take too long. @guybedford: function call parameters are already handled but both external exports and attaching things to globals are missing. Totally forgot about the latter, will include them in the fix.

**Comment by Havunen** [comment]
This is actually still broken in Rollup v60 also

**Comment by lukastaegert** [comment]
Ah yes, I should have checked vs. the original repro. Turns out there was *another* issue at work here in that reassignments are not traced through object literals. Minimal repro:
https://rollupjs.org/repl?version=0.60.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwb2JqJTIwJTNEJTIwJTdCJTVDbiU1Q3R4JTNBJTIwKCklMjAlM0QlM0UlMjAlN0IlN0QlNUNuJTdEJTVDbiU1Q253aW5kb3cueCUyMCUzRCUyMCU3Qm9iaiU3RCUzQiU1Q24lNUNub2JqLngoKSUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

Hope to be able to post a fix soon.

**Comment by lukastaegert** [comment]
Ah, it seems the issue runs a little deeper. Not sure I will be able to post a fix today.

**Comment by lukastaegert** [comment]
Rollup 0.60.5 now handles your case correctly (and is also a little faster when doing it).

**Comment by Havunen** [comment]
Awesome! Thanks for fast response!
